lib/class-wp-rest-sidebars-controller.php

```diff
@@ -0,0 +1,200 @@
+		return rest_ensure_response( $data );
+	}
+
+	public function update_item_permissions_check( $request ) {
+		if ( ! current_user_can( 'edit_theme_options' ) ) {
+			return new WP_Error(
+				'rest_user_cannot_edit',
+				__( 'Sorry, you are not allowed to edit sidebars.', 'gutenberg' )
+			);
+		}
+
+		return true;
+	}
+
+	public function update_item( $request ) {
+		$result = $this->update_sidebar_blocks( $request['id'], $request );
+		if ( is_wp_error( $result ) ) {
+			return $result;
+		}
+
+		$data = $this->get_sidebar_data( $request['id'] );
+		return rest_ensure_response( $data );
+	}
+
+	// TODO: Add schema
+
+	protected function get_sidebar_data( $sidebar_id ) {
+		global $wp_registered_sidebars;
+
+		if ( ! isset( $wp_registered_sidebars[ $sidebar_id ] ) ) {
+			return new WP_Error(
+				'rest_sidebar_invalid_id',
+				__( 'Invalid sidebar ID.', 'gutenberg' ),
+				array( 'status' => 404 )
+			);
+		}
+
+		// TODO: How should we format blocks in the REST API? Or should we send down
+		// HTML so that we're consistent with the /posts and /blocks endpoints?
+		$blocks = array();
+
+		$sidebars_items = gutenberg_get_sidebars_items();
+		if ( ! empty( $sidebars_items[ $sidebar_id ] ) ) {
+			foreach ( $sidebars_items[ $sidebar_id ] as $item ) {
+				if ( is_array( $item ) && isset( $item['blockName'] ) ) {
+					$blocks[] = array(
+						'name'         => $item['blockName'],
+						'attributes'   => $item['attrs'],
+						'innerBlocks'  => $item['innerBlocks'],
+						'innerHTML'    => $item['innerHTML'],
+						'innerContent' => $item['innerContent'],
+					);
+				} else {
+					$blocks[] = array(
+						'name'         => 'core/legacy-widget',
+						'attributes'   => array( 'identifier' => $item ),
+						'innerBlocks'  => array(),
+						'innerHTML'    => '',
+						'innerContent' => array(),
+					);
+				}
+			}
+		}
+
+		return array_merge(
+			$wp_registered_sidebars[ $sidebar_id ],
+			array( 'blocks' => $blocks )
+		);
+	}
+
+	protected function update_sidebar_blocks( $sidebar_id, $request ) {
+		global $wp_registered_sidebars;
+
+		if ( ! isset( $wp_registered_sidebars[ $sidebar_id ] ) ) {
+			return new WP_Error(
+				'rest_sidebar_invalid_id',
+				__( 'Invalid sidebar ID.', 'gutenberg' ),
+				array( 'status' => 404 )
+			);
+		}
+
+		$items = array();
+
+		if ( isset( $request['blocks'] ) && is_array( $request['blocks'] ) ) {
+			foreach ( $request['blocks'] as $block ) {
+				if (
+					! isset( $block['name'] ) || ! is_string( $block['name'] ) ||
+					! isset( $block['attributes'] ) || ! is_array( $block['attributes'] ) ||
+					! isset( $block['innerBlocks'] ) || ! is_array( $block['innerBlocks'] ) ||
+					! isset( $block['innerHTML'] ) || ! is_string( $block['innerHTML'] ) ||
+					! isset( $block['innerContent'] ) || ! is_array( $block['innerContent'] )
+				) {
+					continue;
+				}
+
+				if ( 'core/legacy-widget' === $block['name'] ) {
+					$items[] = $block['attributes']['identifier'];
```

Comment:
Yeah, this isn't quite correct. I think this speaks to my comment below:

> In this PR I assumed that core/legacy-widget is responsible for updating the original widget. Your comment in https://github.com/WordPress/gutenberg/pull/13511/files#r261867151 has made me realise that this isn't correct, though.
> 
> I'll need to do some experimenting, but I am thinking that, when a legacy widget is updated, we store the updates in the core/legacy-widget block. This should be backwards compatible because rendering a core/legacy-widget delegates rendering to the_widget. It does mean that plugins that rely on directly reading a widget's options will stop working, though. We will also need to find a solution for updating callback widgets.